code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -0,0 +1,90 @@
+use alloc::string::{String, ToString};
+use sp_std::{marker::PhantomData, str::FromStr};
+
+use frame_support::{ensure, RuntimeDebug};
+use ibc::core::{
+	ics04_channel::{
+		channel::{Counterparty, Order},
+		error::Error as IbcError,
+		Version,
+	},
+	ics24_host::identifier::{ChannelId, ConnectionId, PortId},
+	ics26_routing::context::{Module as IbcModule, ModuleId, ModuleOutputBuilder},
+};
+use pallet_ibc::routing::ModuleRouter as IbcModuleRouter;
+
+use crate::{Config, Pallet};
+
+#[derive(Default, RuntimeDebug, Eq, PartialEq, Clone)]
+pub struct Router<T: Config> {
+	_marker: PhantomData<T>,
+}
+
+impl<T: Config + Send + Sync> IbcModule for Router<T> {
+	fn on_chan_open_init(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		order: Order,
+		_connection_hops: &[ConnectionId],
+		port_id: &PortId,
+		_channel_id: &ChannelId,
+		_counterparty: &Counterparty,
+		_version: &Version,
+	) -> Result<(), IbcError> {
+		let port_contract =
+			<Pallet<T>>::cosmwasm_addr_to_account(port_id.to_string()).map_err(|_| {
+				IbcError::implementation_specific("contract for port not found".to_string())
+			})?;
+
+		// this check to be moved into contract
+		ensure!(
+			order == Order::Unordered,
+			IbcError::unknown_order_type("can work only with Unordered channels".to_string())
+		);
+		if let Ok(_contract) = <Pallet<T>>::contract_info(&port_contract) {
+			log::trace!(target: "cosmwasm::ibc", "contract for port found");
+			// here we should call `ibc_channel_open`
+			Ok(())
+		} else {
+			Err(IbcError::implementation_specific(
+				"contract for desired port not found".to_string(),
+			))
+		}
+	}
+
+	fn on_chan_open_try(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		_order: Order,
+		_connection_hops: &[ConnectionId],
+		_port_id: &PortId,
+		_channel_id: &ChannelId,
+		_counterparty: &Counterparty,
+		_version: &Version,
+		_counterparty_version: &Version,
+	) -> Result<Version, IbcError> {
+		// here we should call `ibc_channel_open`
+		Err(IbcError::implementation_specific("unimplemented!".to_string()))
+	}
+}
+
+impl<T: Config + Send + Sync + Default> IbcModuleRouter for Router<T> {
+	fn get_route_mut(
+		&mut self,
+		module_id: &impl core::borrow::Borrow<ModuleId>,
+	) -> Option<&mut dyn IbcModule> {
+		if module_id.borrow() == &ModuleId::from_str("cosmwasm").expect("constant") {
+			return Some(self)
+		}
+
+		None
+	}
+
+	fn has_route(module_id: &impl sp_std::borrow::Borrow<ModuleId>) -> bool {
+		module_id.borrow() == &ModuleId::from_str("cosmwasm").expect("constant")
+	}
+
+	fn lookup_module_by_port(_port_id: &PortId) -> Option<ModuleId> {
```

Comment:
You should check if pallet cosm wasm has a contract assigned to this port before returning the module Id as `cosmwasm`, if not this will overwrite every other module including `ics20`.

Reply:
I work it as part of XCVM. It will use cw20-ics20. So we do not care about the ics20 pallet now. Later refactoring is possible. So ics20 will be contract too. 

cc @hussein-aitlahcen

Reply:
will impement ics20 router as part of https://app.clickup.com/t/3ufp2hr

Reply:
resolve for now. see task tag to review ics20 router enabled (`composite` router, so can test each router independently) and join via type Router = (Router1, Router2)